### ISSUE TITLE
The JPMS boundary should be concerned during access to serializer

### ISSUE THREAD

**Issue by iseki0** [Kotlin/kotlinx.serialization#2429]
**Describe the bug**
When we create a hand-written serializer, it can be linked with some type by `@Serializable` annotation.
The hand-written serializer could be placed to another package which is not exported by `module-info.class`.
The `Json.decodeFromString<Type>(data)` will be failed with an `IllegalAccessException` because of the JPMS edge.

**Expected behavior**
Could we add some warning for it?

**Environment**
 - Kotlin version: 1.8.22
 - Library version: 1.8.22
 - Kotlin platforms: JVM
 - Gradle version: 8.2

**Comment by iseki0** [comment]
Before that, I'm trying use JPMS to hide serializers in Java side. (The internal visibility modifier is not working in Java.)

**Comment by pdvrieze** [comment]
Because serialization is intended to be statically resolved the serializer type is part of the ABI. As such it being (potentially) hidden by JPMS is expected behaviour.

ps. if you want to hide it from Java you can try the `@JvmSynthetic` annotation as tools will ignore synthetic types/members.
pps. there is nothing that prevents serialization to be used from Java or makes it awkward (it isn't like inline classes or coroutines), so exposing the serializer types is valid.

**Comment by iseki0** [comment]
@pdvrieze I means there's no compilation error if the serializaer is hidden. As a instead there're runtime exception thrown during try to serialize/deserialize.

**Comment by iseki0** [comment]
🤣I change the title, make it more clearly

**Comment by pdvrieze** [comment]
@iseki0 I see, I assume that you use the `Type.serializer()` function. This is however an intrinsic now that will silently be replaced with the actual serializer. Your issue is probably with the intrinsic not observing JPMS.

**Comment by iseki0** [comment]
Although I'm not using it.(I'm using the standalone function `serializer<Type>()` instead). But the problem you said is correct. The intrinsic ignored the JPMS boundray during compilation. Sorry for my bad English, I use a wrong word. @pdvrieze

**Comment by sandwwraith** [comment]
Hmm, interesting issue. Shouldn't just using a class reference (e.g. `println(MyFooSerializer::class)`) also trigger a JPMS error? In that case, this is a missing functionality in Kotlin or IDE, rather than a serialization plugin.

**Comment by iseki0** [comment]
@sandwwraith emmm Do you meaning to export a type with an internal serializer is a problem and it should be treat as an error by IDE and Kotlin compiler?

**Comment by sandwwraith** [comment]
No, I mean referencing non-exported type should be an error in the IDE regardless whether it is used in serialization annotations or not.

Or, maybe I misunderstood you and you have a situation like this:

```
// File a.kt
package my

@Serializable(my.internal.MySerializer::class)
class MyClass

// File b.kt
package my.internal
class MySerializer: KSerializer<MyClass>

// module-info.java
exports my;
```

**Comment by iseki0** [comment]
I think the problem is clarificated. :)🥰

**Comment by sandwwraith** [comment]
Related: https://youtrack.jetbrains.com/issue/KT-71680/Report-a-warning-when-Serializable-public-class-has-private-serializer

I think eventually we should migrate to 'call serializer() function by default' if we aren't doing that already.

**Comment by pdvrieze** [comment]
There should not be a need for code to have direct access to the type of the serializer, only obtaining one is needed (for which `serializer()` or `serializer(typeParamSerializer…)` would be sufficient.
Of course this would limit the ability to specify the serializer in `@Serializable(with=MySerializer::class)`, but that already doesn't work with generated serializers.

**Comment by vsartiushkin** [comment]
I was unable to reproduce `IllegalAccessException` with the `Json.decodeFromString<Type>(data)` call. Using both the reported versions Kotlin 1.8.22 with library version 1.5.1 and the latest Kotlin 2.4.10 with library version 1.11.0. 

`IllegalAccessException` only appears when code explicitly passes the serializer class into the runtime lookup. 
The `Json.decodeFromString<Type>(data)` works correctly when the serializer for MyClass is not exported by the module.

I did find a way to make `decodeFromString` produce `IllegalAccessError`. But it can only be done when the serialization plugin is applied to the calling Kotlin module and **not** applied to the module containing the class and its custom serializer. 
However this is not a supported setup - the serialization plugin is required for modules that define `@Serializable` classes.

@iseki0 Can you please provide a sample project that reproduces the `IllegalAccessException` thrown by `decodeFromString` call specifically?

**Comment by iseki0** [comment]
@vsartiushkin 3 years ago, I forgot many thing. Maybe... I guess it's related sometime we cannot generate the `serializer()` method. Such as, maybe we add `@Serializable` annotation to a Java class. In that case the Kotlin compiler cannot generate the method...

**Comment by vsartiushkin** [comment]
@iseki0 Thank you! 

This case does reproduce the `IllegalAccessError` (not 'IllegalAccessException') on `decodeFromString`/`encodeToString` call:
```
// File MyClass.java
package my

@Serializable(with = MySerializer.class)
public class MyClass

// File MySerializer.kt
package my.internal
class MySerializer: KSerializer<MyClass>

// module-info.java
exports my;
```

I reproduced the issue both with the initial reported versions and current Kotlin and compiler plugin 2.4.10 and library version 1.11.0 
Attached configured reproducer project [serialization_gh_2429.zip](https://github.com/user-attachments/files/31219603/serialization_gh_2429.zip)

**Comment by pdvrieze** [comment]
@vsartiushkin In this case `MyClass` is defined in Java (with the annotation provided in Java). This means that `MyClass` behaves as if the plugin was not applied at all (it ignores Java classes). As such, only direct access would be possible.

While the direct access could be resolved at compile time by the consumer plugin, this direct access does require the class to be visible.

Unfortunately giving a warning/error about this is not easy as: 
 - In the producer (`lib`) the serialization compiler plugin does not process the `MyClass.java` file (or its associated class), so it doesn't see anything that it could warn about. It also does not know about JPMS modules.
 - In the consumer (`app`) the serialization plugin can see `MyClass.class` and its annotations, but it does not handle JPMS. As such it does not know the serializer is not visible.

As such, to do something more sensible really requires JPMS support (or some cross-platform system that maps to/from it) in the compiler/compiler plugin. Such would be a larger project out of the scope of kotlinx.serialization.
